### PR TITLE
S3 crawler

### PR DESCRIPTION
**Pull request by usefulalgorithm**
### 🤔 Why?

Adds support for parsing S3 objects as datasets. Specifically, the following scenarios are implemented:
- Parsing files to datasets
- Parsing directories to datasets
    - this is done by finding the newest file within the directory, and using that as the source of truth
    - to do this, specify `{table}` label in the path specification: `s3://.../.../{table}/*/*.csv`
    - It is also possible to specify other labels to be included in the dataset name. For example, `s3://{dept}/*/{name}/*/{table}/*.csv` will have the crawler name all parsed datasets as `DEPT/NAME/TABLE`
- Parsing partitioned tables to datasets
    - specify the path spec as `s3://.../.../{table}/{partition_key[0]}={partition[0]/*.parquet` to parse partitioned tables as datasets. If there's no partition key, then write `s3://.../.../{table}/{partition[0]}/*.parquet`.

### 🤓 What?

- Implementation
- Readme
- Unit test

### 🧪 Tested?

- Added unit test with actual minio storage
- Tested end to end

截圖 2024-01-02 晚上11 08 51

This is done with https://github.com/MetaphorData/app/pull/7485

### ☑️ Checks

- [x] My PR contains actual code changes, and I have updated the version number in `pyproject.toml`.

### CODE REVIEW

File: metaphor/common/logger.py [hunk 1]
```diff
@@ -24,6 +24,7 @@ def get_logger() -> logging.Logger:
     logger = logging.getLogger("metaphor")
     logger.setLevel(logging.DEBUG)
     logger.addHandler(console)
+    console.setLevel(logging.DEBUG)
```
Comment (usefulalgorithm) [inline metaphor/common/logger.py:27]: no

File: metaphor/s3/table_data.py [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+    @property
+    def guid(self) -> str:
+        return self.table_path
+
+    @property
+    def logical_id(self) -> DatasetLogicalID:
+        return DatasetLogicalID(
+            name=yarl.URL(self.table_path)
+            .with_scheme("")
+            .human_repr()[1:],  # Drop the first slash
```
Comment (elic-eon) [inline metaphor/s3/table_data.py:68]: What does this transformation do? Drop the `s3://` and first slash?

File: tests/s3/test_table_data.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+    assert merged == TableData(
+        display_name="foo",
+        full_path="bar",
+        partitions=None,
+        timestamp=datetime(2024, 1, 1, 0, 0, 0, 0),
+        table_path="foo",
+        size_in_bytes=5678,
+        number_of_files=1,
+    )
```
Comment (elic-eon) [inline tests/s3/test_table_data.py:37]: Why merged `TableData` have different `full_path` and `table_path`?

Comment (usefulalgorithm) [reply tests/s3/test_table_data.py:37]: full_path = path to the file
table_path = path to the dataset

a dataset can be a single file or a directory containing other files / directories.

File: metaphor/s3/table_data.py (hunk 2 above)
Comment (usefulalgorithm) [reply metaphor/s3/table_data.py:68]: yes.

File: metaphor/s3/path_spec.py [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
+        return self.url.with_scheme("").human_repr()[1:]  # Drop the first slash
+
+    @cached_property
+    def path_prefix(self) -> str:
+        """
+        Returns the prefix of the path before the label level. E.g.
+        for uri = `s3://foo/bar/baz/*/*/{table}/*/*.*`, this would be
+        `bar/baz/*/*/`.
+        """
+        index = re.search(r"[\*|\{]", self.object_path)
```
[the remark this answers is not in the document]
Comment (usefulalgorithm) [reply metaphor/s3/path_spec.py:127]: the comment was incorrect.

File: metaphor/s3/extractor.py [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+                folder1 = folder1.rsplit("=", 1)[-1]
+                folder2 = folder2.rsplit("=", 1)[-1]
+
+        num_folder1 = int(folder1)
+        num_folder2 = int(folder2)
+        if num_folder1 == num_folder2:
+            return 0
+        else:
+            return 1 if num_folder1 > num_folder2 else -1
+    except Exception:
```
[the remark this answers is not in the document]
Comment (usefulalgorithm) [reply metaphor/s3/extractor.py:34]: `num_folder1 = int(folder1)` can throw

File: metaphor/s3/extractor.py [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+        self._path_specs = config.path_specs
+
+    def _resolve_templated_folders(
+        self, bucket_name: str, path_prefix: str
+    ) -> Iterable[str]:
+        folder_split: List[str] = path_prefix.split("*", 1)
+        # If the len of split is 1 it means we don't have * in the prefix
+        if len(folder_split) == 1:
+            yield path_prefix
+            return
```
[the remark this answers is not in the document]
Comment (usefulalgorithm) [reply metaphor/s3/extractor.py]: actually no, we have to explicitly return after yielding here

File: metaphor/s3/extractor.py [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+        iterator = list_folders(
+            bucket_name=bucket_name,
+            prefix=folder,
+            config=self._config,
+        )
+        iterator = peekable(iterator)
+        if iterator:
+            sorted_dirs = sorted(
+                iterator,
+                key=functools.cmp_to_key(
```
[the remark this answers is not in the document]
Comment (usefulalgorithm) [reply metaphor/s3/extractor.py:100]: can't think of a better way to compare two directories...

[the remark this answers is not in the document]
Comment (elic-eon) [reply metaphor/s3/extractor.py:98]: nvm, use `cmp_to_key` is correct.

File: tests/s3/test_table_data.py (hunk 3 above)
Comment (elic-eon) [reply tests/s3/test_table_data.py:37]: Merging two `TableData` with different `table_path` didn't make sense in the crawler? So this is only for the test case?

My real question is the logic looks different between `test_merge_table_data` and `test_merge_table_data_with_partitions`.

Comment (usefulalgorithm) [reply tests/s3/test_table_data.py:37]: I see what you mean now. Yes, in a real scenario we would never merge two table datas with different table paths. The objects here are only for testing.

I will add more comment in the merge method, see if it makes more sense.

**Review by elic-eon — APPROVED** [review]
LGTM

**Pull request merged**